src/server/failed_logins.rs

```diff
@@ -0,0 +1,188 @@
+use crate::options::FailedLoginsPenalty;
+use crate::options::FailedLoginsPolicy;
+
+use super::shutdown;
+use slog::Logger;
+use std::collections::HashMap;
+use std::net::IpAddr;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::{Mutex, RwLock};
+
+#[derive(Hash, Eq, PartialEq, Debug, Clone)]
+struct FailedLoginsKey {
+    ip: Option<IpAddr>,
+    username: Option<String>,
+}
+
+#[derive(Debug, Clone)]
+struct FailedLoginsEntry {
+    attempts: u32,
+    last_attempt_at: Instant,
+}
+
+impl FailedLoginsEntry {
+    fn new() -> Mutex<FailedLoginsEntry> {
+        Mutex::new(FailedLoginsEntry {
+            attempts: 1,
+            last_attempt_at: Instant::now(),
+        })
+    }
+
+    fn time_elapsed(&self) -> Duration {
+        self.last_attempt_at.elapsed()
+    }
+
+    fn touch(&mut self) {
+        self.last_attempt_at = Instant::now();
+    }
+}
+
+/// Temporarily remembers failed logins
+#[derive(Debug)]
+pub struct FailedLoginsCache {
+    policy: FailedLoginsPolicy,
+    penalty: FailedLoginsPenalty,
+    failed_logins: Arc<RwLock<HashMap<FailedLoginsKey, Mutex<FailedLoginsEntry>>>>,
+}
+
+#[derive(Debug)]
+pub enum FailedLoginsError {
```

Comment:
This is used as an error but it doesn't actually implement the Error Trait.

You can import `use thiserror::Error;` and then just do `#[derive(Debug, Error)]
`

Reply:
Here is an example in our code:

```rust
/// The Error returned by storage backends. Storage backend implementations should choose the
/// `ErrorKind` chosen for errors carefully since that will determine what is returned to the FTP
/// client.
#[derive(Debug, Error)]
#[error("storage error: {kind}")]
pub struct Error {
    kind: ErrorKind,
    #[source]
    source: Option<BoxError>,
}

```